Rewrite the HappyProcessStart to more correctly deal with environment

robszewczyk opened the pull request:
The existing implementation of HappyProcessStart had a number of bugs
in cases where the environment variables contained "complicated"
values, i.e. values containing spaces, quotes, backslashes,
semicolons, etc. This implementation improves on the previous
implementation in that the environment variables can now contain the
above characters and will be properly escaped.

robszewczyk committed:
Rewrite the HappyProcessStart to more correctly deal with environment

The existing implementation of HappyProcessStart had a number of bugs
in cases where the environment variables contained "complicated"
values, i.e. values containing spaces, quotes, backslashes,
semicolons, etc. This implementation improves on the previous
implementation in that the environment variables can now contain the
above characters and will be properly escaped.

robszewczyk commented:
Note to reviewers: this PR addresses the intermittent failures we've observed in `openweave-core` CI.  I have validated that with the PR, the tests that were failing from time to time on CI -- namely, `happy/tests/standalone/pairing/test_weave_pairing_01.py.log` and `happy/tests/standalone/pairing/test_weave_ble_01.py.log` -- pass on my local machine

jenniexie commented on happy/HappyProcessStart.py:
```diff
@@ -261,32 +261,53 @@ def __start_daemon(self):
+            need_bash = True
 
         if need_internal_sudo:
+            tmp = "ls"
```
could you explain why do we change to "ls" command here?

robszewczyk replied:
This is clearly breaking abstractions.  I'm creating a command to be run as superuser (whatever that requires within happy), using a dummy command (`ls`) and then I'm stripping the command itself, and I process the command prefix to parse whatever was necessary to run the command as the requested user into a list.  A better way to do this would have been to add something like `Driver.getRunAsRootPrefixList()` and `Driver.getRunAsUserPrefixList()`

robszewczyk replied:
On reflection, I decided that it was the right time to add those abstractions.  The PR has been modified to reflect that.

robszewczyk committed:
Added abstractions for getting command prefix lists for changing users

Rather than working backwards from `runAsRoot` to extract commands
`sudo` invocations, introduce new abstractions into `Driver.py`:
`getRunAsRootPrefixList` and `getRunAsUserPrefixList` that return a
list of strings that need to be prefixed to the command list in order
to run as a requested user.
